File: ios/Runner/Api/PlayerHostApiImpl.swift
//
//  PlayerHostApiImpl.swift
//  Runner
//
//  Created by tk on 2024/06/25.
//

class PlayerHostApiImpl: PlayerHostApi {
    private let albumRepository = AlbumRepository()
    private let songRepository = SongRepository()

    func start(index: Int64, ids: [String], repeatMode: RepeatMode, shuffleMode: ShuffleMode) throws {
        let songIds = ids.map { SongId(id: $0) }
        let songs = songRepository.findBySongIds(songIds: songIds)

        MusicPlayer.shared.start(list: songs, currentIndex: Int(index), repeatMode: repeatMode, shuffleMode: shuffleMode)
    }

    func startAllSongs(index: Int64, repeatMode: RepeatMode, shuffleMode: ShuffleMode) throws {
        let songs = songRepository.findAll()

        MusicPlayer.shared.start(list: songs, currentIndex: Int(index), repeatMode: repeatMode, shuffleMode: shuffleMode)
    }

    func startAlbumSongs(index: Int64, albumId: String, repeatMode: RepeatMode, shuffleMode: ShuffleMode) throws {
        let songs = songRepository.findByAlbumId(albumId: AlbumId(id: albumId))

        MusicPlayer.shared.start(list: songs, currentIndex: Int(index), repeatMode: repeatMode, shuffleMode: shuffleMode)
    }

    func startArtistSongs(index: Int64, artistId: String, repeatMode: RepeatMode, shuffleMode: ShuffleMode) throws {
        let albums = albumRepository.findByArtistId(artistId: ArtistId(id: artistId))
        let albumIds = albums.map { $0.id }
        let songs = songRepository.findByAlbumIds(albumIds: albumIds)

        MusicPlayer.shared.start(list: songs, currentIndex: Int(index), repeatMode: repeatMode, shuffleMode: shuffleMode)
    }

    func play() throws {
        MusicPlayer.shared.play()
    }
    
    func pause() throws {
        MusicPlayer.shared.pause()
    }
    
    func prev() throws {
        MusicPlayer.shared.prev()
    }

    func next() throws {
        MusicPlayer.shared.next()
    }

    func seek(time: Double) throws {
        MusicPlayer.shared.seek(time: time)
    }

    func setRepeat(repeatMode: RepeatMode) throws {
        MusicPlayer.shared.setRepeat(repeatMode: repeatMode)
    }

    func setShuffle(shuffleMode: ShuffleMode) throws {
        MusicPlayer.shared.setShuffle(shuffleMode: shuffleMode)
    }

    func getCurrentTime() throws -> Double {
        return MusicPlayer.shared.getCurrentTime()
    }
}

File: android/app/src/main/kotlin/dev/tcode/thinmpf/api/PlayerHostApiImpl.kt
package dev.tcode.thinmpf.api

import android.content.Context
import dev.tcode.thinmpf.model.valueObject.AlbumId
import dev.tcode.thinmpf.model.valueObject.ArtistId
import dev.tcode.thinmpf.model.valueObject.SongId
import dev.tcode.thinmpf.pigeon_output.PlayerHostApi
import dev.tcode.thinmpf.pigeon_output.RepeatMode
import dev.tcode.thinmpf.pigeon_output.ShuffleMode
import dev.tcode.thinmpf.player.MusicPlayer
import dev.tcode.thinmpf.repository.SongRepository

class PlayerHostApiImpl(private val context: Context): PlayerHostApi {
    override fun start(index: Long, ids: List<String>, repeatMode: RepeatMode, shuffleMode: ShuffleMode) {
        val repository = SongRepository(context)
        val songIds = ids.map { SongId(it) }
        val songs = repository.findByIds(songIds)

        MusicPlayer.start(songs, index.toInt(), repeatMode, shuffleMode)
    }

    override fun startAllSongs(index: Long, repeatMode: RepeatMode, shuffleMode: ShuffleMode) {
        val repository = SongRepository(context)
        val songs = repository.findAll()

        MusicPlayer.start(songs, index.toInt(), repeatMode, shuffleMode)
    }

    override fun startAlbumSongs(index: Long, albumId: String, repeatMode: RepeatMode, shuffleMode: ShuffleMode) {
        val repository = SongRepository(context)
        val songs = repository.findByAlbumId(AlbumId(albumId))

        MusicPlayer.start(songs, index.toInt(), repeatMode, shuffleMode)
    }

    override fun startArtistSongs(index: Long, artistId: String, repeatMode: RepeatMode, shuffleMode: ShuffleMode) {
        val repository = SongRepository(context)
        val songs = repository.findByArtistId(ArtistId(artistId))

        MusicPlayer.start(songs, index.toInt(), repeatMode, shuffleMode)
    }

    override fun play() {
        MusicPlayer.play()
    }

    override fun pause() {
        MusicPlayer.pause()
    }

    override fun prev() {
        MusicPlayer.prev()
    }

    override fun next() {
        MusicPlayer.next()
    }

    override fun seek(time: Double) {
        MusicPlayer.seek(time.toLong() * 1000)
    }

    override fun setRepeat(repeatMode: RepeatMode) {
        MusicPlayer.setRepeat(repeatMode)
    }

    override fun setShuffle(shuffleMode: ShuffleMode) {
        MusicPlayer.setShuffle(shuffleMode)
    }

    override fun getCurrentTime(): Double {
        return MusicPlayer.getCurrentTime().toDouble() / 1000
    }
}